pkg/controller/hpaenforcement_controller.go

```diff
@@ -0,0 +1,555 @@
+		Named(HPAEnforcementCtrlName).
+		Watches(
+			&source.Kind{Type: &v1alpha1.PolicyRecommendation{}},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(predicate.And(predicate.ResourceVersionChangedPredicate{}, updatePredicate, namespaceFilter)),
+		).
+		Watches(
+			&source.Kind{Type: &kedaapi.ScaledObject{}},
+			handler.EnqueueRequestsFromMapFunc(enqueueFunc),
+			builder.WithPredicates(deletePredicate),
+		).
+		Watches(
+			&source.Kind{Type: &v1.Deployment{}},
+			handler.EnqueueRequestsFromMapFunc(policyrecoEnqueueFunc),
+			builder.WithPredicates(predicate.Or(predicate.GenerationChangedPredicate{}, predicate.AnnotationChangedPredicate{})),
+		).
+		Watches(
+			&source.Kind{Type: &argov1alpha1.Rollout{}},
+			handler.EnqueueRequestsFromMapFunc(policyrecoEnqueueFunc),
+			builder.WithPredicates(predicate.Or(predicate.GenerationChangedPredicate{}, predicate.AnnotationChangedPredicate{})),
+		).
+		WithEventFilter(namespaceFilter).
+		Complete(r)
+}
+
+func (r *HPAEnforcementController) isWhitelistedNamespace(namespace string) bool {
+	if r.IncludedNamespaces != nil && len(*r.IncludedNamespaces) > 0 {
+		for _, ns := range *r.IncludedNamespaces {
+			if namespace == ns {
+				return true
+			}
+		}
+		return false
+	}
+
+	fmt.Printf("Excluded ns list : %v\n", r.ExcludedNamespaces)
+	if r.ExcludedNamespaces != nil && len(*r.ExcludedNamespaces) > 0 {
+		for _, ns := range *r.ExcludedNamespaces {
+			if namespace == ns {
+				fmt.Println("Namespace in blacklist.")
+				return false
+			}
+		}
+		fmt.Println("Namespace not in blacklist.")
+		return true
+	}
+
+	fmt.Println("Namespace not in blacklist or whitelist.")
+	return true
+}
+
+func (r *HPAEnforcementController) deleteControllerManagedScaledObject(ctx context.Context, policyreco v1alpha1.PolicyRecommendation, logger logr.Logger) error {
+	//TODO: instead set min = max ? Not correct since in some cases there shouldn't be a scaledobject
+	scaledObjects := &kedaapi.ScaledObjectList{}
+	labelSelector, err := labels.Parse(fmt.Sprintf("%s=%s", createdByLabelKey, createdByLabelValue))
+	if err != nil {
+		logger.V(0).Error(err, "Unable to parse label selector string.")
+		return err
+	}
+	// List only scaledObjects not created by this controller
+	if err := r.List(ctx, scaledObjects, &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(scaledObjectField, policyreco.GetName()),
+		LabelSelector: labelSelector,
+		Namespace:     policyreco.GetNamespace(),
+	}); err != nil && client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	for _, scaledObject := range scaledObjects.Items {
+		r.Delete(ctx, &scaledObject)
+		logger.V(0).Info("Deleted ScaledObject for the policyreco.", "policyreco.name", policyreco.GetName(), "policyreco.namespace", policyreco.GetNamespace(), "scaledobject.name", scaledObject.Name, "scaledobject.namespace", scaledObject.Namespace)
+	}
+
+	//TODO: reset the spec.replicas to the value from the max-pods annotation
```

Comment:
If it is not there, do we want to reset spec.Replicas from the scaledobjects maxReplicas?